Implemented scrolling disabling for ScrollView via ScrollOrientation.Neither

AxelUser opened the pull request:
### Description of Change ###

- Implemented scrolling disabling for `ScrollView` via `ScrollOrientation.Neither`.

### Issues Resolved ###

- fixes #2680 

### API Changes ###

Changed:
- Added `Neither` to `ScrollOrientation`;
- Added handling of `Neither` option in `ScrollView`;

### Platforms Affected ###

<!-- Please list all platforms affected by these changes -->

- Core/XAML (all platforms)

### Behavioral/Visual Changes ###

Nothing will change for old code, because these fixes handle only cases for `ScrollView` with `Orientation = ScrollOrientation.Neither`

### PR Checklist ###

- [x] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=3113) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=3113) before we can accept your contribution.<br/><br/>:x: AxelUser [sign now](https://cla.dotnetfoundation.org/xamarin/Xamarin.Forms?pullRequest=3113)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/xamarin/Xamarin.Forms?pullRequest=3113) it.</sub>

samhouts commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue2680ScrollView.cs:
```diff
@@ -0,0 +1,47 @@
+			RunningApp.ScrollDown(); 
+			RunningApp.ScrollDown(); 
+			RunningApp.WaitForElement(q => q.Marked("Test label 1")); 
+			RunningApp.Screenshot("I'm still seeing the first Label in non-scrolable ScrollView"); 
```
non-scrollable >> non-scrollable

AxelUser replied:
Fixed in https://github.com/xamarin/Xamarin.Forms/pull/3113/commits/9538045d025c61f6861c89caba30ae4c275707eb

PureWeen commented:
@AxelUser  can you rebase and fix the merge conflicts on this one? I realize it's been awhile so just let me know if you won't have time

AxelUser commented:
Hi, @PureWeen! I've fixed conflicts now, is it ok?

samhouts commented:
@AxelUser Just pinging to see if you're able to make the requested changes. Thanks!

AxelUser commented:
Hi, @samhouts! Yes, I'll do changes in tests in couple of days!

rmarinho commented on Xamarin.Forms.Core/ScrollView.cs:
```diff
@@ -228,6 +232,10 @@ protected override SizeRequest OnSizeRequest(double widthConstraint, double heig
 					widthConstraint = double.PositiveInfinity;
 					heightConstraint = double.PositiveInfinity;
 					break;
+				case ScrollOrientation.Neither:
```
Can we add a unit test ?

AxelUser replied:
Wow, I think that would be great

AxelUser replied:
@rmarinho I've started updating unit-test in a9a89f848b32dc899251f8f0be9ac9dc11279874, but I feel a little bit lost. May you assist me in what test cases should I cover?

rmarinho replied:
need to test the size is the same with Neither, and maybe that the scrolled even't doesn't fire?

AxelUser replied:
@rmarinho  Thank you for recomendations! I've added tests'n'fixed in 0128f0033c7593222477e661b69d0fa7d96f5549 and 445bc9208d87b3dcd7657e28cc6f12126d113926. Btw, that helped to fixe firing scroll event when its not necessary!

rmarinho commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue2680ScrollView.cs:
```diff
@@ -0,0 +1,47 @@
+#if UITEST
+		[Test] 
+		public void Issue2680Test() 
+		{ 
```
Can we add another test to show changing between the orientation like @samhouts said?

AxelUser replied:
Yes, that will help a lot!

AxelUser replied:
@rmarinho I've added ui-test in 52ba97934bd783c9a34be9e828350989ac1d0eee

samhouts commented:
@AxelUser Would you be able to address the comments above? Thanks!

AxelUser commented:
@rmarinho @samhouts do you need something else? Did I missed something?

samhouts commented:
@AxelUser Would you be able to add the tests that @rmarinho mentioned?

charlesroddie commented:
@AxelUser are the previous requests feasible for you:
> Please add a test that shows you can switch between Orientations at runtime (e.g., add a button that changes the Orientation to Both).
> Also, please remove WIP from the title when you're ready for this to be merged.
> Can we add a unit test and improve UITest? thanks

charlesroddie commented:
@AxelUser Why is there another PR on this issue? https://github.com/xamarin/Xamarin.Forms/pull/3062

AxelUser commented:
@samhouts @rmarinho thank you for ideas and patience, I've implemented requested changes, what's next?

rmarinho commented:
/AzurePipelines run

rmarinho commented:
Hi @AxelUser you mind rebase ? Thanks

charlesroddie commented:
@AxelUser In this PR, will setting ScrollOrientation.Neither also stop the handling of touch input by the ScrollView? (Our interest in this PR is to fix platform-specific custom slider inside scrollview issues by setting ScrollOrientation.Neither temporarily (https://github.com/halkar/xamarin-range-slider/issues/101).)

AxelUser commented:
@charlesroddie I'll check it out

AxelUser commented:
@charlesroddie I've tried to run example from issue, but couldn't reproduce problem on my laptop, I don't have any Windows devices with touch screen.
@rmarinho I've rebased and squashed commits. Did your CI agents produce some nugets as artifacts, so @charlesroddie could target them at his test project?

AxelUser commented:
@samhouts @rmarinho Hi! Are there any updates?

AxelUser commented:
@samhouts @rmarinho Hi! Are there any updates?

AxelUser commented:
@samhouts Hi, I've mentioned that some checks for iOS 12 have failed, should I do something? Also I don't have access to pipeline, so I can't review fails myself.

StephaneDelcroix commented on Xamarin.Forms.Core/ScrollOrientation.cs:
```diff
@@ -4,6 +4,7 @@ public enum ScrollOrientation
 		Horizontal,
-		Both
+		Both,
+		Neither
```
I'd go for `None`

StephaneDelcroix replied:
/cc @samhouts

AxelUser replied:
Hm, I'm not absolutely sure, but I think that "Neither" is more accurate antonym for "Both".
https://english.stackexchange.com/questions/50693/what-is-the-difference-between-none-of-and-neither-of

samhouts replied:
I like `Neither`. We set the stage for the silly naming with `Both`. For better or worse, it now reads like a question in English. "Which orientation should this scroll: Vertical, Horizontal, Both, or Neither?" 

If instead of `Both`, we had used `Bidirectional` or something, then I would have fully supported `None`.

AxelUser replied:
@StephaneDelcroix do you agree on `Neither`?

AxelUser commented:
Hi, @rmarinho! I'm facing problems while trying to run `Xamarin.Forms.ControlGallery.Android`.
```
The "ResolveLibraryProjectImports" task failed unexpectedly.
System.IO.FileNotFoundException: Could not load assembly 'Xamarin.Forms.Material, Version=0.0.0.0, Culture=neutral, PublicKeyToken='. Perhaps it doesn't exist in the Mono for Android profile?
File name: 'Xamarin.Forms.Material.dll'
   at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.Resolve(AssemblyNameReference reference, ReaderParameters parameters)
   at Java.Interop.Tools.Cecil.DirectoryAssemblyResolver.Resolve(String fullName)
   at Xamarin.Android.Tasks.ResolveLibraryProjectImports.Extract(DirectoryAssemblyResolver res, ICollection`1 jars, ICollection`1 resolvedResourceDirectories, ICollection`1 resolvedAssetDirectories, ICollection`1 resolvedEnvironments)
   at Xamarin.Android.Tasks.ResolveLibraryProjectImports.Execute()
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext()	Embedding.Droid			
```

Have dealt with that issue? Do you have any stacktraces or logs from failed test run?

rmarinho commented:
@AxelUser I think maybe it didn't restore packages correctly ? are you using a Mac ?

samhouts commented:
@AxelUser Make sure you're opening the correct solution file. Don't open the one starting with a dot (.).

AxelUser commented:
@samhouts thank you for making a point about sln file!

@rmarinho I've manually test view with disabled scroll, with/without fast-renderers, but nothing crashed. Do you have more details about UI test results? Or maybe you can somehow share access to pipeline?

charlesroddie commented:
Can this be merged now @StephaneDelcroix @rmarinho @samhouts ?

jfversluis commented:
Unfortunately there is a build error

> Xamarin.Forms.Controls.Issues\Xamarin.Forms.Controls.Issues.Shared\Issue2680ScrollView.cs(20,15): Error CS0108: 'Issue2680ScrollView.IsEnabled' hides inherited member 'VisualElement.IsEnabled'. Use the new keyword if hiding was intended.

charlesroddie commented:
@AxelUser

AxelUser commented:
@jfversluis thank you! I've fixed error, may you restart pipelines?

jfversluis commented:
The UI Test for this issue on Android is failing. iOS seems OK, other failures seem unrelated.

### Issue2680 Test Scroll Enabled

> System.Exception : Error while performing WaitForNoElement(Marked("Test label 1"), "Timed out waiting for no element...", null, null, null)  ----> System.TimeoutException : Timed out waiting for no element...  at Xamarin.UITest.Utils.ErrorReporting.With (System.Action func, System.Object[] args, System.String memberName) [0x00052] in <2a16c16730a54859bda72c6bc1c728f7>:0   at Xamarin.UITest.Android.AndroidApp.WaitForNoElement (System.Func`2[T,TResult] query, System.String timeoutMessage, System.Nullable`1[T] timeout, System.Nullable`1[T] retryFrequency, System.Nullable`1[T] postTimeout) [0x00032] in <2a16c16730a54859bda72c6bc1c728f7>:0   at Xamarin.Forms.Controls.ScreenshotConditionalApp.WaitForNoElement (System.Func`2[T,TResult] query, System.String timeoutMessage, System.Nullable`1[T] timeout, System.Nullable`1[T] retryFrequency, System.Nullable`1[T] postTimeout) [0x00001] in <c8108639f1894d8191ce463e44620a6d>:0   at Xamarin.Forms.Controls.Issues.Issue2680ScrollView.Issue2680Test_ScrollEnabled () [0x00073] in <c8108639f1894d8191ce463e44620a6d>:0   at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in <48b95f3df5804531818f80e28ec60191>:0 --TimeoutException  at Xamarin.UITest.Shared.WaitForHelper.WaitFor (System.Func`1[TResult] predicate, System.String timeoutMessage, System.Nullable`1[T] timeout, System.Nullable`1[T] retryFrequency, System.Nullable`1[T] postTimeout) [0x0002a] in <2a16c16730a54859bda72c6bc1c728f7>:0   at Xamarin.UITest.Android.AndroidApp+<>c__DisplayClass87_0.<WaitForNoElement>b__0 () [0x00054] in <2a16c16730a54859bda72c6bc1c728f7>:0   at Xamarin.UITest.Utils.ErrorReporting.With (System.Action func, System.Object[] args, System.String memberName) [0x00006] in <2a16c16730a54859bda72c6bc1c728f7>:0

AxelUser commented:
@jfversluis how do you run UI-tests for issues on local device?

AxelUser commented:
@jfversluis I've fixed UI tests, may you check pipelines?

AxelUser commented:
@jfversluis I saw that some checks on android failed, may you send an error message?

AxelUser commented:
Thank you, I've fixed codestyle

jfversluis commented:
Thank you for your contribution!

rmalmoe commented:
Is there an ETA on when this will be released?

jfversluis commented:
This got merged into master which means at this point it's planned for 4.3. If I look at the roadmap here: https://github.com/xamarin/Xamarin.Forms/wiki/Feature-Roadmap 4.3 is scheduled for Sept/Oct. Before that there should probably be a pre-release.

Hope this gives you the information you need

rmalmoe commented:
Hello @jfversluis,

Thank you for the update! We will workaround this until the release.

Miksier commented:
Whenever I set orientation as neither the scroll indeed gets disabled, but is also squezes entire page into the space as if the ScrollView never existed. 
From what I understand this should only turn off scrolling of ScrollView not recalculate the layout while removing the ScrollView expandable view.

redradist commented:
> 
> 
> Whenever I set orientation as neither the scroll indeed gets disabled, but is also squeezes entire page into the space as if the ScrollView never existed.
> From what I understand this should only turn off scrolling of ScrollView not recalculate the layout while removing the ScrollView expandable view.
> 
> Ok, from code I see that its doing exactly that, but that way its useless. It does not disable scrolling at all. Now if I want to DISABLE scroll in the middle of the page(lets say 3x device height) the content just gets squeezed into 1 page.
> Now scroll disabling must still be done in custom renderers

Also faced with the same issue ... By some reason ScrollView became some strange Layout ... I also thought it should disable scrolling without recalculating whole View ?!

redradist commented:
Also strange that after page was rendered it is not possible to disable scrolling ... !!

@AxelUser @samhouts Is it intended behavior ? This feature works only on first rendering ?
Is it possible to disable scrolling for some time and then enable it without ScrollView disabled as View ?

AxelUser commented:
@redradist hm, actually it was possible to disable scrolling dynamically.
